File: app/src/main/java/tw/y_studio/ptt/UI/BaseFragment.java
package tw.y_studio.ptt.UI;

import android.app.Activity;
import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;

import androidx.fragment.app.Fragment;

import tw.y_studio.ptt.Utils.DebugUtils;

public class BaseFragment extends Fragment {

    protected Context mContext = null;
    protected BaseActivity mActivity = null;
    protected View mMainView = null;
    private Handler mUIHandler = new Handler(Looper.getMainLooper());

    public void setActivity(BaseActivity activity){
        this.mActivity = activity;
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        if(activity instanceof BaseActivity){
            this.mActivity = (BaseActivity)activity;
        }



    }

    public <T extends View> T findViewById(int id){
       return (T) mMainView.findViewById(id);
    }

    protected void onAnimOver(){

    }

    protected void runOnUI(Runnable r){
        mUIHandler.post(r);
    }

    protected void setMainView(View view){
        this.mMainView = view;
    }

    protected View getMainView(){
        return this.mMainView;
    }

    private boolean isFirstStart = false;
    @Override
    public Animation onCreateAnimation(int transit, final boolean enter, int nextAnim) {

        if(isFirstStart) return super.onCreateAnimation(transit,enter,nextAnim);

        try{
            Animation anim = AnimationUtils.loadAnimation(mActivity, nextAnim);

            anim.setAnimationListener(new Animation.AnimationListener() {

                public void onAnimationStart(Animation animation) {

                }

                public void onAnimationRepeat(Animation animation) {

                }

                public void onAnimationEnd(Animation animation) {

                    isFirstStart=true;

                    onAnimOver();
                    animation.setAnimationListener(null);

                }
            });
            return anim;
        }catch (Exception e){
            onAnimOver();
            isFirstStart=true;
            return super.onCreateAnimation(transit,enter,nextAnim);
        }
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.mContext = context;

    }

    @Override
    public void onDetach() {
        super.onDetach();
        this.mContext = null;
        this.mActivity = null;
        this.mMainView = null;

    }

    @Override
    public void onDestroy(){
        super.onDestroy();
    }

    @Override
    public Context getContext(){
        if(this.mContext == null){
            return super.getContext();
        }else {
            return this.mContext;
        }
    }

    public Activity getCurrentActivity(){
        if(this.mActivity != null){
            return this.mActivity;
        }else if(this.mContext != null){
            return (Activity) this.mContext;
        }else{
            return getActivity();
        }
    }

    public void closeAllFragment(){
        mActivity.closeAllFragment();
    }

    protected Fragment getCurrentFragment(){
        return (Fragment) this;
    }
    public void loadFragment(Fragment toFragment, Fragment thisFragment){

        try {
            mActivity.loadFragment(toFragment,thisFragment);
        } catch (Exception e) {
            e.printStackTrace();
            DebugUtils.Log("loadFragment","Error : "+e.getLocalizedMessage());
        }

    }

    public void loadFragmentNoAnim(Fragment toFragment, Fragment thisFragment){

        try {
            mActivity.loadFragmentNoAnim(toFragment,thisFragment);
        } catch (Exception e) {
            e.printStackTrace();
            DebugUtils.Log("loadFragmentNoAnim","Error : "+e.getLocalizedMessage());
        }

    }
}

File: app/src/main/java/tw/y_studio/ptt/ui/BaseFragment.kt
package tw.y_studio.ptt.ui

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import androidx.fragment.app.Fragment
import tw.y_studio.ptt.utils.Log

open class BaseFragment : Fragment() {

    private var _mContext: Context? = null
    protected val mContext get() = _mContext!!

    private var mActivity: BaseActivity? = null

    private var _mainView: View? = null
    protected val mainView get() = _mainView!!

    private val mUIHandler = Handler(Looper.getMainLooper())

    fun setActivity(activity: BaseActivity?) {
        mActivity = activity
    }

    fun setMainView(view: View) {
        _mainView = view
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return super.onCreateView(inflater, container, savedInstanceState).apply {
            _mainView = this
        }
    }

    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        if (activity is BaseActivity) {
            mActivity = activity
        }
    }

    fun <T : View?> findViewById(id: Int): T {
        return mainView.findViewById<View>(id) as T
    }

    protected open fun onAnimOver() {}
    protected fun runOnUI(r: Runnable) {
        mUIHandler.post(r)
    }

    private var isFirstStart = false
    override fun onCreateAnimation(transit: Int, enter: Boolean, nextAnim: Int): Animation? {
        return if (isFirstStart) super.onCreateAnimation(transit, enter, nextAnim) else try {
            val anim = AnimationUtils.loadAnimation(mActivity, nextAnim)
            anim.setAnimationListener(
                object : Animation.AnimationListener {
                    override fun onAnimationStart(animation: Animation) {}
                    override fun onAnimationRepeat(animation: Animation) {}
                    override fun onAnimationEnd(animation: Animation) {
                        isFirstStart = true
                        onAnimOver()
                        animation.setAnimationListener(null)
                    }
                })
            anim
        } catch (e: Exception) {
            onAnimOver()
            isFirstStart = true
            super.onCreateAnimation(transit, enter, nextAnim)
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        _mContext = context
    }

    override fun onDetach() {
        super.onDetach()
        _mContext = null
        mActivity = null
        _mainView = null
    }

    override fun getContext(): Context {
        return if (_mContext != null) {
            _mContext!!
        } else if (mActivity != null) {
            mActivity!!
        } else {
            mainView.context!!
        }
    }

    val currentActivity: Activity
        get() = requireActivity()

    fun closeAllFragment() {
        mActivity?.closeAllFragment()
    }

    protected val currentFragment: Fragment get() = this

    fun loadFragment(toFragment: Fragment?, thisFragment: Fragment?) {
        try {
            mActivity?.loadFragment(toFragment, thisFragment)
        } catch (e: Exception) {
            e.printStackTrace()
            Log("loadFragment", "Error : " + e.localizedMessage)
        }
    }

    fun loadFragmentNoAnim(toFragment: Fragment?, thisFragment: Fragment?) {
        try {
            mActivity?.loadFragmentNoAnim(toFragment, thisFragment)
        } catch (e: Exception) {
            e.printStackTrace()
            Log("loadFragmentNoAnim", "Error : " + e.localizedMessage)
        }
    }
}
